File: ai/src/services/approval.service.ts
/**
 * F5 — Final Submission Approval & Archival
 */
import pool from '../config/db';
import { getAutoApproveRules } from '../engine/thresholds';
import type { ApprovalCheck } from '../types/ai.types';

export const checkAutoApproval = async (projectId: number): Promise<ApprovalCheck> => {
    const rules = getAutoApproveRules();
    const reasons: string[] = [];
    const projRes = await pool.query(`SELECT id, progress, status, review_state FROM projects WHERE id = $1`, [projectId]);
    if (projRes.rows.length === 0) {
        return { project_id: projectId, eligible: false, reasons: ['Project not found'],
            criteria: { progress_complete: false, all_tasks_done: false, min_docs_approved: false, topic_approved: false, no_pending_revisions: false }};
    }
    const project = projRes.rows[0];

    const progress_complete = project.progress >= rules.min_progress;
    if (!progress_complete) reasons.push(`Progress ${project.progress}%, need ${rules.min_progress}%`);

    const taskRes = await pool.query(`SELECT COUNT(*)::int AS total, COUNT(*) FILTER (WHERE status='done')::int AS done FROM tasks WHERE project_id=$1`, [projectId]);
    const all_tasks_done = taskRes.rows[0]?.total === 0 || taskRes.rows[0]?.done === taskRes.rows[0]?.total;
    if (!all_tasks_done) reasons.push(`${taskRes.rows[0]?.total - taskRes.rows[0]?.done} task(s) incomplete`);

    const docRes = await pool.query(`SELECT COUNT(*)::int AS approved FROM documents WHERE project_id=$1 AND status='Approved'`, [projectId]);
    const min_docs_approved = (docRes.rows[0]?.approved || 0) >= rules.min_docs_approved;
    if (!min_docs_approved) reasons.push(`${docRes.rows[0]?.approved||0} doc(s) approved, need ${rules.min_docs_approved}`);

    const topic_approved = project.review_state === 'Approved';
    if (!topic_approved) reasons.push(`Topic state "${project.review_state}", needs "Approved"`);

    const pendingRes = await pool.query(`SELECT COUNT(*)::int AS count FROM documents WHERE project_id=$1 AND status IN ('Pending','Rejected')`, [projectId]);
    const no_pending_revisions = (pendingRes.rows[0]?.count || 0) === 0;
    if (!no_pending_revisions) reasons.push(`${pendingRes.rows[0]?.count} doc(s) pending/rejected`);

    const eligible = progress_complete && all_tasks_done && min_docs_approved && topic_approved && no_pending_revisions;
    return { project_id: projectId, eligible, reasons: eligible ? ['All criteria met'] : reasons,
        criteria: { progress_complete, all_tasks_done, min_docs_approved, topic_approved, no_pending_revisions }};
};

export const batchCheckApprovals = async (batchId?: number) => {
    const filter = batchId ? 'AND b.id = $1' : '';
    const params = batchId ? [batchId] : [];
    const res = await pool.query(`SELECT p.id FROM projects p JOIN groups g ON g.id=p.group_id LEFT JOIN batches b ON b.id=g.batch_id WHERE p.status!='rejected' ${filter}`, params);
    const results: ApprovalCheck[] = [];
    for (const row of res.rows) results.push(await checkAutoApproval(row.id));
    return { total_checked: results.length, eligible: results.filter(r => r.eligible), not_eligible: results.filter(r => !r.eligible) };
};

export const autoApproveProject = async (projectId: number) => {
    const check = await checkAutoApproval(projectId);
    if (!check.eligible) return { success: false, message: 'Not eligible', reasons: check.reasons };
    await pool.query(`UPDATE projects SET status='approved', updated_at=NOW() WHERE id=$1`, [projectId]);
    await pool.query(`INSERT INTO ai_audit_log (action, entity_type, entity_id, details) VALUES ('auto_approve','project',$1,$2)`, [projectId, JSON.stringify(check.criteria)]);
    return { success: true, message: 'Project auto-approved', project_id: projectId };
};

export const archiveCompletedProjects = async (batchId: number) => {
    const res = await pool.query(`UPDATE projects p SET review_state='Archived', updated_at=NOW() FROM groups g WHERE g.id=p.group_id AND g.batch_id=$1 AND p.status='approved' AND p.progress=100 AND p.review_state!='Archived' RETURNING p.id, p.title`, [batchId]);
    if (res.rows.length > 0) await pool.query(`INSERT INTO ai_audit_log (action, details) VALUES ('archive_projects', $1)`, [JSON.stringify({ batch_id: batchId, archived: res.rows })]);
    return { archived_count: res.rows.length, projects: res.rows };
};

export const getArchiveSummary = async (batchId: number) => {
    const res = await pool.query(`SELECT COUNT(*)::int AS total, COUNT(*) FILTER (WHERE review_state='Archived')::int AS archived, COUNT(*) FILTER (WHERE status='approved' AND review_state!='Archived')::int AS completed, COUNT(*) FILTER (WHERE status='pending')::int AS in_progress FROM projects p JOIN groups g ON g.id=p.group_id WHERE g.batch_id=$1`, [batchId]);
    return res.rows[0];
};

File: ai-py/services/approval_service.py
"""
F5 — Final Submission Approval & Archival
"""
import json
from config.database import query, execute
from config.settings import get_auto_approve_rules


def check_auto_approval(project_id: int) -> dict:
    rules = get_auto_approve_rules()
    reasons = []
    proj = query("SELECT id, progress, status, review_state FROM projects WHERE id = %s", (project_id,))
    if not proj:
        return {"project_id": project_id, "eligible": False, "reasons": ["Project not found"],
                "criteria": {k: False for k in ["progress_complete", "all_tasks_done", "min_docs_approved", "topic_approved", "no_pending_revisions"]}}
    p = proj[0]

    progress_complete = (p["progress"] or 0) >= rules["min_progress"]
    if not progress_complete:
        reasons.append(f"Progress {p['progress']}%, need {rules['min_progress']}%")

    tasks = query("SELECT COUNT(*)::int AS total, COUNT(*) FILTER (WHERE status='done')::int AS done FROM tasks WHERE project_id = %s", (project_id,))
    t = tasks[0] if tasks else {"total": 0, "done": 0}
    all_tasks_done = t["total"] == 0 or t["done"] == t["total"]
    if not all_tasks_done:
        reasons.append(f"{t['total'] - t['done']} task(s) incomplete")

    docs = query("SELECT COUNT(*)::int AS approved FROM documents WHERE project_id = %s AND status = 'Approved'", (project_id,))
    approved_cnt = docs[0]["approved"] if docs else 0
    min_docs_approved = approved_cnt >= rules["min_docs_approved"]
    if not min_docs_approved:
        reasons.append(f"{approved_cnt} doc(s) approved, need {rules['min_docs_approved']}")

    topic_approved = p["review_state"] == "Approved"
    if not topic_approved:
        reasons.append(f'Topic state "{p["review_state"]}", needs "Approved"')

    pend = query("SELECT COUNT(*)::int AS count FROM documents WHERE project_id = %s AND status IN ('Pending','Rejected')", (project_id,))
    pend_cnt = pend[0]["count"] if pend else 0
    no_pending_revisions = pend_cnt == 0
    if not no_pending_revisions:
        reasons.append(f"{pend_cnt} doc(s) pending/rejected")

    eligible = all([progress_complete, all_tasks_done, min_docs_approved, topic_approved, no_pending_revisions])
    return {
        "project_id": project_id, "eligible": eligible,
        "reasons": ["All criteria met"] if eligible else reasons,
        "criteria": {"progress_complete": progress_complete, "all_tasks_done": all_tasks_done,
                     "min_docs_approved": min_docs_approved, "topic_approved": topic_approved,
                     "no_pending_revisions": no_pending_revisions},
    }


def batch_check_approvals(batch_id: int = None) -> dict:
    filt = "AND b.id = %s" if batch_id else ""
    params = (batch_id,) if batch_id else None
    rows = query(
        f"""SELECT p.id FROM projects p JOIN groups g ON g.id = p.group_id
            LEFT JOIN batches b ON b.id = g.batch_id WHERE p.status != 'rejected' {filt}""",
        params,
    )
    results = [check_auto_approval(r["id"]) for r in rows]
    return {"total_checked": len(results),
            "eligible": [r for r in results if r["eligible"]],
            "not_eligible": [r for r in results if not r["eligible"]]}


def auto_approve_project(project_id: int) -> dict:
    check = check_auto_approval(project_id)
    if not check["eligible"]:
        return {"success": False, "message": "Not eligible", "reasons": check["reasons"]}
    execute("UPDATE projects SET status = 'approved', updated_at = NOW() WHERE id = %s", (project_id,))
    execute("INSERT INTO ai_audit_log (action, entity_type, entity_id, details) VALUES ('auto_approve','project',%s,%s)",
            (project_id, json.dumps(check["criteria"])))
    return {"success": True, "message": "Project auto-approved", "project_id": project_id}


def archive_completed_projects(batch_id: int) -> dict:
    rows = execute(
        """UPDATE projects p SET review_state = 'Archived', updated_at = NOW()
           FROM groups g WHERE g.id = p.group_id AND g.batch_id = %s
           AND p.status = 'approved' AND p.progress = 100 AND p.review_state != 'Archived'
           RETURNING p.id, p.title""",
        (batch_id,),
    )
    if rows:
        execute("INSERT INTO ai_audit_log (action, details) VALUES ('archive_projects', %s)",
                (json.dumps({"batch_id": batch_id, "archived": rows}),))
    return {"archived_count": len(rows), "projects": rows}


def get_archive_summary(batch_id: int):
    rows = query(
        """SELECT COUNT(*)::int AS total,
                  COUNT(*) FILTER (WHERE review_state='Archived')::int AS archived,
                  COUNT(*) FILTER (WHERE status='approved' AND review_state!='Archived')::int AS completed,
                  COUNT(*) FILTER (WHERE status='pending')::int AS in_progress
           FROM projects p JOIN groups g ON g.id = p.group_id WHERE g.batch_id = %s""",
        (batch_id,),
    )
    return rows[0] if rows else {}
